File: src/VanEngine.Core/VAN/JuulLexer.cs
using VanEngine.Core.VAN;

namespace VanEngine.Core.VAN;

public ref struct JuulLexer
{
    private ReadOnlySpan<char> _input;
    private int _position;

    public JuulLexer(ReadOnlySpan<char> input)
    {
        _input = input;
        _position = 0;
    }

    public JuulMask? ReadNextMask()
    {
        while (_position < _input.Length && char.IsWhiteSpace(_input[_position]))
            _position++;

        if (_position >= _input.Length)
            return null;

        char c = _input[_position];
        _position++;

        return FryasAlphabet.GetMask(c);
    }

    public JuulMask[] ToMaskArray()
    {
        var list = new List<JuulMask>();
        while (ReadNextMask() is { } mask)
            list.Add(mask);
        return list.ToArray();
    }

    public int Position => _position;
    public bool HasMore => _position < _input.Length;
}

File: ConversationIDE/src/VanEngine.Core/VAN/juul_lexer.py
from __future__ import annotations

from .fryas_alphabet import FryasAlphabet
from .juul_mask import JuulMask


class JuulLexer:
    def __init__(self, input_text: str):
        self._input = input_text
        self._position = 0

    def read_next_mask(self) -> JuulMask | None:
        while self._position < len(self._input) and self._input[self._position].isspace():
            self._position += 1

        if self._position >= len(self._input):
            return None

        ch = self._input[self._position]
        self._position += 1
        return FryasAlphabet.get_mask(ch)

    def to_mask_array(self) -> list[JuulMask]:
        masks: list[JuulMask] = []
        while True:
            m = self.read_next_mask()
            if m is None:
                break
            masks.append(m)
        return masks

    @property
    def position(self) -> int:
        return self._position

    @property
    def has_more(self) -> bool:
        return self._position < len(self._input)
